### DISCUSSION TITLE
How to keep the Attachments path when uploading to the Github Repository?

### DISCUSSION THREAD

**Discussion by carlos-truong** [ObsidianPublisher/obsidian-github-publisher#311] [Q&A — Plugin] [answered]
Hi,

I create new attachments folder for each of my note. For example, Note1's attachments will be put in "Attachments/Note1" and Note2's attachments will be put in "Attachments/Note2". 

But after uploading my notes to the Github Repository by Obsidian Publisher,  all of the attachments will be put in "Attachments". But I want to put the attachments of each note into separate folders.
Is there any way to do it?

**Comment by Mara-Li** [comment]
Nope, but you can use the frontmatter to set a custom folder for each note. 
If you want this function, create a FR on the repository.

**Reply by Mara-Li** [reply, marked as answer]
Note: It's not possible to allow that, because it's not possible to know which file is connected to which note. Plus it cause duplicate image. 

Moreover, if you use this type of things in Obsidian, you can use the Obsidian structure in Embeds → Structure (click on enable)

**Comment by carlos-truong** [comment]
Hi Lisandra,
Thank you for your reply!

I've tried both these two ways but it doesn't work.

1) My vault:
  MyVault/MyBlog/Attachments/Category1/image1.png
  MyVault/MyBlog/Attachments/Category1/image2.png
  MyVault/MyBlog/Category1/index.md
    - ![[../Attachments/Category1/image1.png]]
    - ![[../Attachments/Category1/image2.png]]

2) My vault:
  MyVault/MyBlog/Category1/Attachments/image1.png
  MyVault/MyBlog/Category1/Attachments/image2.png
  MyVault/MyBlog/Category1/index.md
    - ![[Attachments/image1.png]]
    - ![[Attachments/image2.png]]

But I received this error when uploading "Category1/index.md" to the repo with these both two ways:
"Unable to upload note image1.png, skipping it"
"Unable to upload note image2.png, skipping it"

I've enabled "Attachments & embeds -> Structure"
And this is my plugin settings:
```json
{
  "github": {
    "branch": "main",
    "automaticallyMergePR": true,
    "dryRun": {
      "enable": false,
      "folderName": "github-publisher"
    },
    "tokenPath": "%configDir%/plugins/%pluginID%/env",
    "api": {
      "tiersForApi": "Github Free/Pro/Team (default)",
      "hostname": ""
    },
    "workflow": {
      "commitMessage": "[PUBLISHER] Merge",
      "name": ""
    },
    "verifiedRepo": true
  },
  "upload": {
    "behavior": "obsidian",
    "defaultName": "docs",
    "rootFolder": "",
    "yamlFolderKey": "",
    "frontmatterTitle": {
      "enable": false,
      "key": "title"
    },
    "replaceTitle": [],
    "replacePath": [
      {
        "regex": "MyBlog",
        "replacement": "",
        "type": "path"
      }
    ],
    "autoclean": {
      "enable": true,
      "excluded": [
        "docs/assets",
        "docs/Attachments",
        "index.md",
        "about.md",
        "tags.md"
      ]
    },
    "folderNote": {
      "enable": false,
      "rename": "index.md",
      "addTitle": {
        "enable": false,
        "key": "title"
      }
    },
    "metadataExtractorPath": ""
  },
  "conversion": {
    "hardbreak": false,
    "dataview": true,
    "censorText": [],
    "tags": {
      "inline": false,
      "exclude": [],
      "fields": []
    },
    "links": {
      "internal": false,
      "unshared": false,
      "wiki": false,
      "slugify": false
    }
  },
  "embed": {
    "attachments": true,
    "overrideAttachments": [],
    "keySendFile": [],
    "notes": false,
    "folder": "docs/Attachments",
    "convertEmbedToLinks": "keep",
    "charConvert": "->",
    "useObsidianFolder": true
  },
  "plugin": {
    "shareKey": "share",
    "excludedFolder": [],
    "copyLink": {
      "enable": false,
      "links": "",
      "removePart": [],
      "transform": {
        "toUri": true,
        "slugify": "lower",
        "applyRegex": []
      }
    },
    "setFrontmatterKey": "Set"
  }
}
```

**Comment by Mara-Li** [comment]
Could you please open the console log with the developer panel?

**Comment by carlos-truong** [comment]
Hi, there are a lot of errors:

<img width="848" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/29748838/7b3ee21f-7b32-4d88-b496-3727db6c9042">

**Reply by Mara-Li** [reply]
I need the response of the 422 error, could you open it more?
I don't have any problem on my side while sending file…

**Comment by carlos-truong** [comment]
Okay, here it is:
```
plugin:obsidian-mkdocs-publisher:85     PUT https://api.github.com/repos/carloslab-ai/carloslab-ai.github.io/contents/docs%2F%2FAttachments%2FUtilityIntelligence%2FAttachments%2Fcreate-utility-world-owner.png 422
D0 @ plugin:obsidian-mkdocs-publisher:85
m @ plugin:obsidian-mkdocs-publisher:85
vS @ plugin:obsidian-mkdocs-publisher:87
eval @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
Jy @ plugin:obsidian-mkdocs-publisher:2
Hg.Object.assign.endpoint @ plugin:obsidian-mkdocs-publisher:85
eval @ plugin:obsidian-mkdocs-publisher:108
f @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
eval @ plugin:obsidian-mkdocs-publisher:2
Be @ plugin:obsidian-mkdocs-publisher:2
upload @ plugin:obsidian-mkdocs-publisher:108
eval @ plugin:obsidian-mkdocs-publisher:108
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
eval @ plugin:obsidian-mkdocs-publisher:2
Be @ plugin:obsidian-mkdocs-publisher:2
uploadImage @ plugin:obsidian-mkdocs-publisher:108
eval @ plugin:obsidian-mkdocs-publisher:108
f @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
f @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
f @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
f @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
eval @ plugin:obsidian-mkdocs-publisher:2
Be @ plugin:obsidian-mkdocs-publisher:2
statusBarForEmbed @ plugin:obsidian-mkdocs-publisher:108
eval @ plugin:obsidian-mkdocs-publisher:108
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
eval @ plugin:obsidian-mkdocs-publisher:2
Be @ plugin:obsidian-mkdocs-publisher:2
uploadOnMultipleRepo @ plugin:obsidian-mkdocs-publisher:108
eval @ plugin:obsidian-mkdocs-publisher:108
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
eval @ plugin:obsidian-mkdocs-publisher:2
Be @ plugin:obsidian-mkdocs-publisher:2
publish @ plugin:obsidian-mkdocs-publisher:108
eval @ plugin:obsidian-mkdocs-publisher:92
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
p @ plugin:obsidian-mkdocs-publisher:2
Promise.then (async)
m @ plugin:obsidian-mkdocs-publisher:2
eval @ plugin:obsidian-mkdocs-publisher:2
Be @ plugin:obsidian-mkdocs-publisher:2
Kc @ plugin:obsidian-mkdocs-publisher:92
checkCallback @ plugin:obsidian-mkdocs-publisher:92
Oq @ app.js:1
t.onChooseItem @ app.js:1
t.onChooseSuggestion @ app.js:1
t.selectSuggestion @ app.js:1
e.useSelectedItem @ app.js:1
(anonymous) @ app.js:1
e.handleKey @ app.js:1
e.onKeyEvent @ app.js:1
```

**Reply by Mara-Li** [reply]
Well. It give me nothing... Sorry. I don't know what cause this error :/

**Reply by carlos-truong** [reply]
How about this?

<img width="541" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/29748838/0fa0f635-d798-4198-b4b7-c65ab7257539">

**Reply by Mara-Li** [reply]
Nan, I don't have any specific error for this. 422 is a error that can be caused by a lot of things : https://docs.github.com/en/rest/using-the-rest-api/troubleshooting-the-rest-api?apiVersion=2022-11-28#validation-failed

I think it's an issue on your repository or configuration

**Reply by carlos-truong** [reply]
I don't know why but I've created a new repository to test it. And I received the same error: 422

**Reply by carlos-truong** [reply]
Could I ask how you configure your repo and your vault for the attachments with "Structure" enabled?
Here is mine:

<img width="512" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/29748838/f755289d-b8d0-4389-bfad-a4faa09dde72">
<img width="608" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/29748838/64157238-7d21-4940-b9cb-808d948d30e5">

**Reply by carlos-truong** [reply]
Hi Lisandra,
I've found that it's due to the Folder Path replacer.
<img width="424" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/29748838/aaf0c187-e47f-4064-ae2b-ec4a05a72d38">

If I remove the Folder Path replacer, everything is okay. So I think there is a bug about replacing the Folder Path when uploading Attachments with Obsidian folder structure.

**Reply by Mara-Li** [reply]
Thanks. Could you please open an issue with a resume of your investigation?

**Reply by carlos-truong** [reply]
Ok, sure!
Here it is:
https://github.com/ObsidianPublisher/obsidian-github-publisher/issues/314
